java/source/data.rst

```diff
@@ -0,0 +1,360 @@
+
+Sorting by manipulating the original vector.
+Algorithm: org.apache.arrow.algorithm.sort.FixedWidthInPlaceVectorSorter - O(nlog(n))
+
+.. testcode::
+
+    import org.apache.arrow.algorithm.sort.DefaultVectorComparators;
+    import org.apache.arrow.algorithm.sort.FixedWidthInPlaceVectorSorter;
+    import org.apache.arrow.algorithm.sort.VectorValueComparator;
+    import org.apache.arrow.vector.IntVector;
+    import org.apache.arrow.memory.RootAllocator;
+
+    RootAllocator rootAllocator = new RootAllocator(Long.MAX_VALUE);
+    IntVector intVectorNotSorted = new IntVector("intvectornotsorted", rootAllocator);
+    intVectorNotSorted.allocateNew(3);
+    intVectorNotSorted.setValueCount(3);
+    intVectorNotSorted.set(0, 10);
+    intVectorNotSorted.set(1, 8);
+    intVectorNotSorted.setNull(2);
+    FixedWidthInPlaceVectorSorter<IntVector> sorter = new FixedWidthInPlaceVectorSorter<IntVector>();
+    VectorValueComparator<IntVector> comparator = DefaultVectorComparators.createDefaultComparator(intVectorNotSorted);
+    sorter.sortInPlace(intVectorNotSorted, comparator);
+
+    System.out.println(intVectorNotSorted);
+
+.. testoutput::
+
+    [null, 8, 10]
+
+Out-place Sorter - O(nlog(n))
+*****************************
+
+Sorting by copies vector elements to a new vector in sorted order - O(nlog(n))
+Algorithm: : org.apache.arrow.algorithm.sort.FixedWidthInPlaceVectorSorter.
+FixedWidthOutOfPlaceVectorSorter & VariableWidthOutOfPlaceVectorSor
+
+.. testcode::
+
+    import org.apache.arrow.algorithm.sort.DefaultVectorComparators;
+    import org.apache.arrow.algorithm.sort.FixedWidthOutOfPlaceVectorSorter;
+    import org.apache.arrow.algorithm.sort.OutOfPlaceVectorSorter;
+    import org.apache.arrow.algorithm.sort.VectorValueComparator;
+    import org.apache.arrow.vector.IntVector;
+    import org.apache.arrow.memory.RootAllocator;
+
+    RootAllocator rootAllocator = new RootAllocator(Long.MAX_VALUE);
+    IntVector intVectorNotSorted = new IntVector("intvectornotsorted", rootAllocator);
+    intVectorNotSorted.allocateNew(3);
+    intVectorNotSorted.setValueCount(3);
+    intVectorNotSorted.set(0, 10);
+    intVectorNotSorted.set(1, 8);
+    intVectorNotSorted.setNull(2);
+    OutOfPlaceVectorSorter<IntVector> sorterOutOfPlaceSorter = new FixedWidthOutOfPlaceVectorSorter<>();
+    VectorValueComparator<IntVector> comparatorOutOfPlaceSorter = DefaultVectorComparators.createDefaultComparator(intVectorNotSorted);
+    IntVector intVectorSorted = (IntVector) intVectorNotSorted.getField().getFieldType().createNewSingleVector("new-out-of-place-sorter", rootAllocator, null);
+    intVectorSorted.allocateNew(intVectorNotSorted.getValueCount());
+    intVectorSorted.setValueCount(intVectorNotSorted.getValueCount());
+    sorterOutOfPlaceSorter.sortOutOfPlace(intVectorNotSorted, intVectorSorted, comparatorOutOfPlaceSorter);
+
+    System.out.println(intVectorSorted);
+
+.. testoutput::
+
+    [null, 8, 10]
+
+Data Filter & Aggregation
```

Comment:
I question the purpose of this recipe as it doesn't show any Arrow capability, I mean it ends up implementing custom filtering and aggregation. We should probably postpone this recipe until Arrow Java gets support for pieces of the compute engine like filtering.

Reply:
Deleted